src/views/primary-auth/PrimaryAuthForm.js

```diff
@@ -38,26 +38,37 @@ define([
     initialize: function () {
       this.listenTo(this, 'save', function () {
         var processCreds = this.settings.get('processCreds');
-        if (_.isFunction(processCreds)) {
-          processCreds({
-            username: this.model.get('username'),
-            password: this.model.get('password')
-          });
+        var asyncProcessCreds = this.settings.get('asyncProcessCreds');
+        var self = this;
+        if(_.isFunction(asyncProcessCreds)) {
+            asyncProcessCreds({
+              username: this.model.get('username'),
+              password: this.model.get('password')
+            }, function(){ postProcessCreds(); });
+        }else{
+          if (_.isFunction(processCreds)) {
+            processCreds({
+              username: this.model.get('username'),
+              password: this.model.get('password')
+            });
+          }
+          postProcessCreds();
         }
-        if (this.settings.get('features.deviceFingerprinting')) {
-          var self = this;
-          DeviceFingerprint.generateDeviceFingerprint(this.settings.get('baseUrl'), this.$el)
-          .then(function (fingerprint) {
-            self.model.set('deviceFingerprint', fingerprint);
-          })
-          .fail(function () {
-            // Keep going even if device fingerprint fails
-          })
-          .fin(function () {
+        function postProcessCreds() {
+          if (this.settings.get('features.deviceFingerprinting')) {
```

Comment:
It doesn't look like "this" is bound to the correct scope here. 

In general, now that theres a bit more logic (including with DeviceFingerprint), I think we can simplify the callbacks by wrapping this in a promise chain:

```javascript
var self = this;

// Note: Will need to import 'q' as a module up top in the AMD definition
Q.Promise(function (resolve) {
  var processCreds = self.settings.get('processCreds');
  var creds = {
    username: self.model.get('username'),
    password: self.model.get('password')
  };

  if (!_.isFunction(processCreds)) {
    return;
  }

  if (processCreds.length === 2) {
    processCreds(creds, resolve);
  } else {
    processCreds(creds);
    resolve();
  }
})
.then(function () {
  if (!self.settings.get('features.deviceFingerprinting')) {
    return;
  }
  return DeviceFingerprint.generateDeviceFingerprint(self.settings.get('baseUrl'), self.$el)
  .then(function (fingerprint) {
    self.model.set('deviceFingerprint', fingerprint);
  })
  .fail(function () {
    // Keep going even if device fingerprint fails
  });
})
.then(function () {
  self.model.save();
});
```